### PR TITLE
Scan merging in spectral library match

### PR DESCRIPTION
**Pull request by robinschmid**
Better handling of MS1 MS2 MSn all or merged spectra

We now have one parameter for MS level + merged or all - instead of two parameters for MS level and All MS2

image

image

### CODE REVIEW

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/RowsSpectralMatchTask.java [hunk 1]
```diff
@@ -467,18 +479,18 @@ private SpectralSimilarity matchSpectrum(Float rowRT, double rowMZ, Float rowCCS
...
         query = cropped[1];
       }
 
       // remove precursor signals
-      if (msLevel > 1 && removePrecursor && ident.getPrecursorMZ() != null) {
+      if (!msLevelFilter.isMs1Only() && removePrecursor && ident.getPrecursorMZ() != null) {
         // precursor mz from library entry for signal filtering
         double precursorMZ = ident.getPrecursorMZ();
         // remove from both spectra
         library = removePrecursor(library, precursorMZ);
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/RowsSpectralMatchTask.java:493]: comment still says remove from both spectra. where did the removal from query spectrum go?

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java [hunk 2]
```diff
@@ -53,14 +55,13 @@ public class SpectralLibrarySearchParameters extends SimpleParameterSet {
...
-      2, 1, 1000);
-
-  public static final BooleanParameter allMS2Spectra = new BooleanParameter(
-      "Check all scans (only for MS2)",
-      "Check all (or only most intense) MS2 scan. This option does not apply to MS1 scans.", false);
-
+  public static final ComboParameter<ScanMatchingSelection> scanMatchingSelection = new ComboParameter<>(
+      "Scans for matching", """
+      Choose the MS level and experimental scans to match against the library. MS1 for GC-EI-MS data,
+      MERGED: will merge all scans, creating one merged spectrum for each fragmentation energy,
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java:61]: will merge all fragment scans.
I think we should reflect this in the scan definition of the merged spectrum

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java [hunk 3]
```diff
@@ -84,55 +85,80 @@ public class SpectralLibrarySearchParameters extends SimpleParameterSet {
...
+
+    @Override
+    public String toString() {
+      return switch (this) {
+        case MS1 -> "MS1";
+        case MERGED_MS2 -> "MS2 (merged)";
+        case MERGED_MSN -> "MS level ≥ 2 (merged)";
+        case ALL_MS2 -> "MS2 (all scans)";
+        case ALL_MSN -> "MS level ≥ 2 (all scans)";
+      };
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java:146]: all of these should be explained in the description

**Review by SteffenHeu — DISMISSED** [review]

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/RowsSpectralMatchTask.java (hunk 1 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/RowsSpectralMatchTask.java:493]: its done when we prepare the spectrum

Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/RowsSpectralMatchTask.java:493]: we extract the mass list etc and now we also do the precursor removal so its only done once

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java (hunk 3 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java:146]: added text on MS2 and MSn

Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java:146]: Choose the MS level and experimental scans to match against the library. MS1 for GC-EI-MS data,
      MERGED: will merge all fragment scans, creating one merged spectrum for each fragmentation energy,
              and one consensus spectrum merged from those different energies.
      ALL: will use all available raw fragment scans + the ones from merging.
      MS2: limits the final list to MS2 scans
      MS2 (merged): and a scan were all MSn scans are merged into one 'pseudo' MS2 scan
      MSn: defines all fragment scans of MS level 2 and higher

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java (hunk 2 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java:61]: we have the MergingType in MergedMassSpectrum. We could add this to the scan definition

Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java:61]: maybe separate PR

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/RowsSpectralMatchTask.java (hunk 1 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/RowsSpectralMatchTask.java:493]: adapt the comment then

File: src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java (hunk 2 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/modules/dataprocessing/id_spectral_library_match/SpectralLibrarySearchParameters.java]: yeah

**Pull request merged**